File: apps/menu-bar/modules/rudder/src/RudderModule.ts
import { requireNativeModule } from 'expo-modules-core';
import { Platform } from 'react-native';

import { RudderClient, NativeRudderModule } from './Rudder.types';

const NativeRudder = requireNativeModule<NativeRudderModule>('Rudder');

const RudderModule: RudderClient = {
  async load(writeKey: string, dataPlaneUrl: string): Promise<void> {
    NativeRudder.load(writeKey, dataPlaneUrl);
  },
  async track(event: string, properties?: Record<string, any>) {
    NativeRudder.track(
      event,
      {
        ...properties,
      },
      {
        os: {
          name: Platform.OS,
          version: Platform.Version,
        },
        app: {
          name: 'orbit',
          version: NativeRudder.appVersion,
          type: 'native',
        },
      }
    );
  },
};

export default RudderModule;

File: apps/menu-bar/modules/rudder/ios/RudderModule.swift
import ExpoModulesCore
import Rudder

public class RudderModule: Module {

  public func definition() -> ModuleDefinition {
    Name("Rudder")
    
    Constants([
      "appVersion": self.appContext?.constants?.constants()["nativeAppVersion"],
    ])

    AsyncFunction("load") { (writeKey: String, dataPlaneUrl:String) in
      let config = RSConfig(writeKey: writeKey)
        .dataPlaneURL(dataPlaneUrl)
        .trackLifecycleEvents(false)

      RSClient.sharedInstance().configure(with: config)
    }.runOnQueue(.main)

    AsyncFunction("track") { (event: String, properties: [String: Any], context: [String: [String: Any]]?) in
      let option = RSOption()
      
      if let context = context, !context.isEmpty {
        for (key, value) in context { 
          option.putCustomContext(value, withKey: key)
        }
      }
      
      RSClient.sharedInstance().track(event, properties:properties, option: option)
    }
  }
}
